File: javascript/packages/linter/src/rules/turbo-permanent-no-misleading-value.ts
import { AttributeVisitorMixin } from "./rule-utils.js"
import { IdentityPrinter } from "@herb-tools/printer"
import { ParserRule, BaseAutofixContext, Mutable } from "../types.js"

import { getStaticAttributeValue, hasAttributeValue, isERBOpenTagNode } from "@herb-tools/core"

import type { StaticAttributeStaticValueParams, StaticAttributeDynamicValueParams } from "./rule-utils.js"
import type { UnboundLintOffense, LintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLAttributeNode, HTMLOpenTagNode, ERBOpenTagNode, ParseResult, ParserOptions } from "@herb-tools/core"

interface TurboPermanentAutofixContext extends BaseAutofixContext {
  node: Mutable<HTMLAttributeNode>
}

class TurboPermanentNoMisleadingValueVisitor extends AttributeVisitorMixin<TurboPermanentAutofixContext> {
  protected checkStaticAttributeStaticValue({ attributeName, attributeNode, parentNode }: StaticAttributeStaticValueParams): void {
    this.checkTurboPermanentAttribute(attributeName, attributeNode, parentNode)
  }

  protected checkStaticAttributeDynamicValue({ attributeName, attributeNode, parentNode }: StaticAttributeDynamicValueParams): void {
    if (isERBOpenTagNode(parentNode)) return

    this.checkTurboPermanentAttribute(attributeName, attributeNode, parentNode)
  }

  private checkTurboPermanentAttribute(attributeName: string, attributeNode: HTMLAttributeNode, parentNode: HTMLOpenTagNode | ERBOpenTagNode): void {
    if (attributeName !== "data-turbo-permanent") return
    if (!hasAttributeValue(attributeNode)) return

    const autofixContext = isERBOpenTagNode(parentNode) ? undefined : { node: attributeNode }
    const isTruthyValue = getStaticAttributeValue(attributeNode)?.trim().toLowerCase() === "true"

    const explanation = isTruthyValue
      ? `is redundant, because Turbo only checks whether the attribute is present.`
      : `still makes the element permanent, because Turbo only checks whether the attribute is present.`

    this.addOffense(
      `Attribute \`data-turbo-permanent\` should not have a value. \`${IdentityPrinter.print(attributeNode)}\` ${explanation} Use \`data-turbo-permanent\` instead.`,
      attributeNode.value!.location,
      autofixContext
    )
  }
}

export class TurboPermanentNoMisleadingValueRule extends ParserRule<TurboPermanentAutofixContext> {
  static autocorrectable = true
  static ruleName = "turbo-permanent-no-misleading-value"
  static introducedIn = this.version("unreleased")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense<TurboPermanentAutofixContext>[] {
    const visitor = new TurboPermanentNoMisleadingValueVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }

  autofix(offense: LintOffense<TurboPermanentAutofixContext>, result: ParseResult, _context?: Partial<LintContext>): ParseResult | null {
    if (!offense.autofixContext) return null

    const { node } = offense.autofixContext

    node.equals = null
    node.value = null

    return result
  }
}

File: rust/herb-linter/src/rules/turbo_permanent_no_misleading_value.rs
use crate::autofix::{for_each_attribute_mut, location_matches};
use crate::offense::Offense;
use crate::rule::LintContext;
use crate::utils::tag_utils::{
  get_attribute_name, get_attributes, get_static_attribute_value, has_attribute_value, has_dynamic_attribute_value, print_attribute,
};

use herb::nodes::{AnyNode, DocumentNode, ERBOpenTagNode, HTMLAttributeNode, HTMLOpenTagNode};
use herb::Visitor;

rule_visitor!(TurboPermanentNoMisleadingValueVisitor);
define_parser_rule!(
  TurboPermanentNoMisleadingValueRule,
  "turbo-permanent-no-misleading-value",
  Error,
  TurboPermanentNoMisleadingValueVisitor,
  parser_options: { action_view_helpers: true },
  autocorrectable: true,
  autofix: autofix,
  introduced_in: "unreleased"
);

impl TurboPermanentNoMisleadingValueVisitor {
  fn check_attribute(&mut self, attribute: &HTMLAttributeNode) {
    if get_attribute_name(attribute).as_deref() != Some("data-turbo-permanent") {
      return;
    }

    if !has_attribute_value(attribute) {
      return;
    }

    let value_node = match attribute.value.as_ref() {
      Some(value) => value,
      None => return,
    };

    // ActionView renders nil and dynamic values as `RubyLiteralNode`, which
    // cannot be resolved statically, so those are not misleading values
    if value_node.children.iter().any(|child| matches!(child, AnyNode::RubyLiteralNode(_))) {
      return;
    }

    let static_value = get_static_attribute_value(attribute);

    let value_location = value_node.location.clone();

    let is_truthy = static_value.map(|value| value.trim().to_lowercase() == "true").unwrap_or(false);

    let explanation = if is_truthy {
      "is redundant, because Turbo only checks whether the attribute is present."
    } else {
      "still makes the element permanent, because Turbo only checks whether the attribute is present."
    };

    self.add_offense(
      format!(
        "Attribute `data-turbo-permanent` should not have a value. `{}` {} Use `data-turbo-permanent` instead.",
        print_attribute(attribute),
        explanation
      ),
      value_location,
    );
  }
}

impl Visitor for TurboPermanentNoMisleadingValueVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    for attribute in get_attributes(node) {
      self.check_attribute(attribute);
    }

    self.walk_html_open_tag_node(node);
  }

  fn visit_erb_open_tag_node(&mut self, node: &ERBOpenTagNode) {
    for child in &node.children {
      if let AnyNode::HTMLAttributeNode(attribute) = child {
        if !has_dynamic_attribute_value(attribute) {
          self.check_attribute(attribute);
        }
      }
    }

    self.walk_erb_open_tag_node(node);
  }
}

fn autofix(offense: &Offense, document: &mut DocumentNode, _context: &LintContext) -> bool {
  let mut fixed = false;

  for_each_attribute_mut(document, &mut |attribute| {
    let matches = attribute
      .value
      .as_ref()
      .map(|value| location_matches(&value.location, offense))
      .unwrap_or(false);

    if !matches {
      return;
    }

    attribute.equals = None;
    attribute.value = None;
    fixed = true;
  });

  fixed
}
